junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TempDirectoryPerDeclarationTests.java

```diff
@@ -1264,6 +1293,83 @@ public void close() throws IOException {
 
 	}
 
+	static class FactoryWithAnnotatedElementNameAsPrefixTestCase {
+
+		@TempDir(factory = Factory.class)
+		private Path tempDir1;
+
+		private Path tempDir2;
+
+		@BeforeEach
+		void setUp(@TempDir(factory = Factory.class) Path tempDir2) {
+			this.tempDir2 = tempDir2;
+		}
+
+		@Test
+		void test(@TempDir(factory = Factory.class) Path tempDir3) {
+			assertThat(tempDir1.getFileName()).asString().startsWith("tempDir1");
+			assertThat(tempDir2.getFileName()).asString().startsWith("tempDir2");
+			assertThat(tempDir3.getFileName()).asString().startsWith("tempDir3");
+		}
+
+		private static class Factory implements TempDirFactory {
+
+			@Override
+			public Path createTempDirectory(AnnotatedElementContext elementContext, ExtensionContext extensionContext)
+					throws Exception {
+				return Files.createTempDirectory(getName(elementContext.getAnnotatedElement()));
+			}
+
+			private static String getName(AnnotatedElement element) {
+				return element instanceof Field ? ((Field) element).getName() : ((Parameter) element).getName();
+			}
+
+		}
+
+	}
+
+	static class FactoryWithCustomMetaAnnotationTestCase {
+
+		@CustomTempDir("field")
+		private Path tempDir1;
```

Comment:
I just realized IntelliJ issues a "never assigned" warning here:

<img width="447" alt="Screenshot 2023-07-16 at 11 29 11" src="https://github.com/junit-team/junit5/assets/26772046/298759c6-f65a-46ba-9843-05ad511f2255">

That's because its inspection rule has no knowledge of the new support for meta-annotations.

I'll raise a ticket to JetBrains.

Reply:
[IDEA-325434](https://youtrack.jetbrains.com/issue/IDEA-325434) raised.